### DISCUSSION TITLE
Waveshare SX1262/SX1268 LoRa HAT For Raspberry Pi

### DISCUSSION THREAD

**Discussion by 2NaughtyDogs** [markqvist/Reticulum#526] [Help & Questions] [answered]
Hello, firstly thank you all for such a great product. 
I have used Rnodes using Lillygo boards and they work great. Now however I would like to build an all in one base station using a raspberry Pi and a lora hat, instead of a tethered Rnode
questions:

1. Firstly has anyone had experience using the Waveshare's SX1262/SX1268 LoRa HATs
2. in the rnsd conf, is it set up as a serial interface or a lora interface.

Thank you for your help.
Cheers

**Comment by nathmo** [comment]
This would be fun. 
here is the wiki to the product you are describing : https://www.waveshare.com/wiki/SX1262_868M_LoRa_HAT
as you seem to suggest, the hat work as a serial link using the E22-400T22S which inside use the SX1262.

From my understanding, you would be able to use it as a serial link between two nodes over Reticulum by configuring the interface as a serial-interface (https://reticulum.network/manual/interfaces.html#serial-interface)

But I don't think other RNode/SPI LoRa Transceiver would be able to use it as a transport node on the LoRa serial interface.

It's unclear to me if multiple (more than two) raspberry with that hat could talk in a meshed manner, and it would require testing.

**Comment by nathmo** [comment, marked as answer]
I'm happy to report that it works :)

![image](https://github.com/user-attachments/assets/41070e15-a2b1-4d3a-8abc-52a8c46559af)

![image](https://github.com/user-attachments/assets/a5838198-e763-499c-924a-eb0c8dcb8e89)

```
ubuntu@nodeA:~ $ cat .reticulum/config
# This is the default Reticulum config file.
# You should probably edit it to include any additional,
# interfaces and settings you might need.

# Only the most basic options are included in this default
# configuration. To see a more verbose, and much longer,
# configuration example, you can run the command:
# rnsd --exampleconfig


[reticulum]

  enable_transport = False
  share_instance = Yes
  shared_instance_port = 37428
  instance_control_port = 37429
  panic_on_interface_error = No


[logging]

  loglevel = 6

[interfaces]

[[Serial Interface]]
  type = SerialInterface
  interface_enabled = True

  # Serial port for the device
  port = /dev/ttyUSB0

  # Set the serial baud-rate and other
  # configuration parameters.
  speed = 9600
  databits = 8
  parity = none
  stopbits = 1
```

**Reply by nathmo** [reply]
this is pretty basic setup that works out of the box with no config. but im sure it's possible to tweak it to 1) connect multiple device 2) have a better data rate

**Reply by cobraPA** [reply]
That's interesting... how are the Lora parameters configured for that device when it is connected over serial?

**Reply by nathmo** [reply]
there are two jumper (and it can alternatively be configured using the GPIO) to switch between command mode or transmit mode 
![image](https://github.com/user-attachments/assets/999de744-1ce0-410a-8162-81a6573647a6)

**Reply by nathmo** [reply]
it's an ebyte modem so you dont have acess to all the LoRa parameters. it encapsulate serial data over LoRa (and the config allow you to change the radio bit rate, the serial baud rate, etc) here is the chip : https://www.cdebyte.com/products/E22-400T22S/1 and the datasheet : https://www.cdebyte.com/pdf-down.aspx?id=3108

**Reply by cobraPA** [reply]
yeah, they mention 'custom protocol', so I guess you don't get access to the params like freq, spreading factor, etc , that we use with RNode Lora.

**Comment by nathmo** [comment]
for futur reference : https://files.waveshare.com/upload/a/af/SX1268_LoRa_HAT_SchDoc.pdf

**Comment by liamcottle** [comment]
I'm interested to know how you get on with this if you introduce several nodes in the same configuration, or intermingle them with other RNode on the same frequency/spreading factors etc.

As far as I'm aware, the RNode firmware implements CSMA to avoid other nodes transmitting at the same time, which would cause packet collisions etc. It also implements a packet queue so it can transmit the queued packets when the airtime is available again.

If you run multiple of these devices in serial mode, without any fancy queuing etc, you'll probably find the packets don't get out properly. Keen to know how it goes :)

I've got the same waveshare hat on one of my Pi 4 boards, acting as a Meshtastic gateway. Was also keen to get this setup working for Reticulum.

**Reply by nathmo** [reply]
The Hat is not using Rnode firmware. It just act as a serial over the air device. You can somewhat configure to have multiple hat connected together with some kind of addressing system, but I don't know if Reticulum will be able to support that. Reticulum accept a serial interface between two devices. (so to make sure it works, just configure two hats to make a link between them, if you want a node with 3 links, then you would need three hats on the node and three other hats for the target.)

It's important to undersand that reticulum and Rnode are two different thing. Reticulum is a network stack. it does the job of a routeur. Rnode is a "transport medium driver" it's one of the many interface reticulum support. (and in this setup using the hat there is no Rnode involved.)

**Reply by liamcottle** [reply]
> Reticulum accept a serial interface between two devices

Yep, I'm more so meaning if you had for example, 10x people using Reticulum, each having one of these LoRa hats as a serial interface, you might run into issues where packets are getting lost/dropped, since the LoRa hat will not have any CSMA/queueing features that the RNode firmware has to prevent it's own LoRa chip from transmitting when other people are transmitting.

> It's important to undersand that reticulum and Rnode are two different thing.

Yessir ;)

> Reticulum is a network stack. it does the job of a router. Rnode is a "transport medium driver"

Yep, as far as I remember, when reading the transport code in RNS, it will send a packet immediately to the interface it wants to send through. In the case of an RNode, which uses LoRa, the RNode firmware will receive this packet, and queue it in firmware until the frequency is clear, using a CSMA slope to help mitigate collisions.

Whereas with the serial device, the packet could get to your LoRa hat at the same time as someone else in your radio range, and it won't have queueing, so it'll either collide with someone elses packet on air, (unless the chip supports listen before transmit), but in the case of it having LBT, it may run out of buffer if several packets are sent to the serial device before it can process them. The latter probably not so much an issue with retries, but collisions will be annoying as it could act as DoS by mistake.

I could be overthinking it all here, but I'm keen to hear if this works well over serial as-is with multiple LoRa hat users.

I've got a couple of these hats, just haven't had a chance to test them with RNS yet. Been too busy working on the RNode web flasher and MeshChat ;)

**Reply by nathmo** [reply]
Well, it seems that you know more than I do. And regarding your concern they seem valid, and I don't have enough hat to test it :(

**Reply by 2NaughtyDogs** [reply]
Hello Nathmo, This is amazing work!  Thank you for the time, my understanding is limited but increasing, but this is awesome.  I would like to thank everyone else that has feed into the discussion, love this community.  I have 3 of these devices which I can test.  I am traveling at the moment, will test when home.  Thank you again!

**Comment by antonionardella** [comment]
Thank you for this post, I was able to reuse HAT which is incompatible with another project.
